src/test/java/com/commercetools/paymenttoorderprocessor/jobs/actions/MessageReaderTest.java

```diff
@@ -0,0 +1,180 @@
+public class MessageReaderTest {
+
+    @Configuration
+    static class TestConfig {
+
+        @Bean
+        public BlockingSphereClient client() {
+            return mock(BlockingSphereClient.class);
+        }
+    }
+
+    private static final String MOCKS = "mocks/messageReader/";
+
+    private static final TypeReference<PagedQueryResult<PaymentTransactionStateChangedMessage>> PagedQueryWithMessages =
+            new TypeReference<PagedQueryResult<PaymentTransactionStateChangedMessage>>() {
+            };
+
+    private static final PagedQueryResult<PaymentTransactionStateChangedMessage> emptyMessagesResult =
+            SphereJsonUtils.readObjectFromResource(MOCKS + "messagesResult_empty.json", PagedQueryWithMessages);
+
+    private static final PagedQueryResult<PaymentTransactionStateChangedMessage> firstMessagesResult =
+            SphereJsonUtils.readObjectFromResource(MOCKS + "messagesResult_1.json", PagedQueryWithMessages);
+
+    private static final PagedQueryResult<PaymentTransactionStateChangedMessage> secondMessagesResult =
+            SphereJsonUtils.readObjectFromResource(MOCKS + "messagesResult_2.json", PagedQueryWithMessages);
+
+    @Autowired
+    private MessageReader messageReader;
+
+    @Autowired
+    private TimeStampManager timeStampManager;
+
+    @Autowired
+    private MessageProcessedManager messageProcessedManager;
+
+    @Autowired
+    private BlockingSphereClient client;
+
+    @Test
+    public void read_whenEmpty_returnsNull() {
+
+        when(client.executeBlocking(any())).thenReturn(emptyMessagesResult);
+
+        assertThat(messageReader.read()).isNull();
+    }
+
+    @Test
+    public void read_whenOnePage_returnsResultFromThePage() {
+        when(client.executeBlocking(any())).thenAnswer(a -> {
+            MessageQuery query = a.getArgumentAt(0, MessageQuery.class);
+            if (query.offset() == 0) {
+                return firstMessagesResult;
+            } else {
+                return emptyMessagesResult;
+            }
+        });
+        PaymentTransactionStateChangedMessage firstUnprocessedMessage = messageReader.read();
+        assertThat(firstUnprocessedMessage).isNotNull();
+        assertThat(firstUnprocessedMessage.getId()).isEqualTo("11111111-1111-1111-1111-111111111111");
+
+        PaymentTransactionStateChangedMessage secondUnprocessedMessage = messageReader.read();
+        assertThat(secondUnprocessedMessage).isNotNull();
+        assertThat(secondUnprocessedMessage.getId()).isEqualTo("44444444-4444-4444-4444-444444444444");
+
+        assertThat(messageReader.read()).isNull(); // empty queue after first page
+
+        // sphere client should be called only twice (first and empty result page)
+        // even if we call messageReader.read 3 times, because both results are on the same page first page
+        verify(client, times(2)).executeBlocking(any());
+
+        // no messages processed - last timestamp is not updated
```

Comment:
I think that in  this case we can also update the timestamp to current time...

Reply:
don't think so: then we have to synchronize service with platform project. I want to keep service  independent from runtime environment and depend only on actual CTP project settings.